compiler/rustc_lint/src/non_local_def.rs

```diff
@@ -0,0 +1,141 @@
+        if self.is_in_body > 0 {
+            match item.kind {
+                ItemKind::Impl(impl_) => {
+                    // The RFC states:
+                    //
+                    // > An item nested inside an expression-containing item (through any
+                    // > level of nesting) may not define an impl Trait for Type unless
+                    // > either the **Trait** or the **Type** is also nested inside the
+                    // > same expression-containing item.
+                    //
+                    // To achieve this we get try to get the paths of the _Trait_ and
+                    // _Type_, and we look inside thoses paths to try a find in one
+                    // of them a type whose parent is the same as the impl definition.
+                    //
+                    // If that's the case this means that this impl block decleration
+                    // is using local items and so we don't lint on it.
+
+                    let parent_impl = cx.tcx.parent(item.owner_id.def_id.into());
+
+                    let self_ty_has_local_parent = if let TyKind::Path(QPath::Resolved(
+                        _,
+                        self_ty_path,
+                    )) = impl_.self_ty.kind
+                    {
+                        has_local_parent(self_ty_path, cx, parent_impl)
+                    } else {
+                        false
+                    };
+
+                    let of_trait_has_local_parent = if let Some(of_trait) = impl_.of_trait {
+                        has_local_parent(of_trait.path, cx, parent_impl)
+                    } else {
+                        false
+                    };
+
+                    // If none of them have a local parent (LOGICAL NOR) this means that
+                    // this impl definition is a non-local definition and so we lint on it.
+                    if !(self_ty_has_local_parent || of_trait_has_local_parent) {
+                        cx.emit_span_lint(
+                            NON_LOCAL_DEFINITIONS,
+                            item.span,
+                            NonLocalDefinitionsDiag { depth: self.is_in_body },
+                        );
+                    }
+                }
+                ItemKind::Macro(_macro, MacroKind::Bang) => {
+                    if cx.tcx.has_attr(item.owner_id.def_id, sym::macro_export) {
+                        cx.emit_span_lint(
+                            NON_LOCAL_DEFINITIONS,
+                            item.span,
+                            NonLocalDefinitionsDiag { depth: self.is_in_body },
+                        );
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+/// Given a path and a parent impl def id, this function tries to find if any of
+/// the segments, generic type argument have a parent def id that correspond to
+/// the def id of the parent impl definition.
```

Comment:
```suggestion
/// Given a path and a parent impl def id, this function tries to find if any of
/// the segments and generic type argument have a parent def id that correspond to
/// the def id of the parent impl definition.
```
I'm unsure what "parent impl definition" and "impl def id" mean though...